transcom/mymove
# INT B-20882 rejected sit resubmission for all sit service items

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20882)

## Summary

During HomeSafe demo testing of SIT, they received an error when resubmitting a `DDFSIT` SIT service item that looked like this:
```
{
  "title": "Validation Error",
  "instance": "66aa4d5f-2e71-9ad1-d185-d1166e3a1d06",
  "detail": "reServiceCode in body should be one of [DDDSIT DDASIT DOPSIT DOASIT DOFSIT]"
}
```
This is a swagger error because we had the enum set up to only allow for those values in the `reServiceCode` key in the payload object.

The payload they sent looked like this:
```
{
  "modelType": "UpdateMTOServiceItemSIT",
  "reServiceCode": "DDFSIT",
  "dateOfContact1": "2024-08-02",
  "dateOfContact2": "2024-08-05",
  "timeMilitary1": "1400Z",
  "timeMilitary2": "1400Z",
  "firstAvailableDeliveryDate1": "2024-08-03",
  "firstAvailableDeliveryDate2": "2024-08-06",
  "sitEntryDate": "2024-08-01",
  "reason": "Boxes too big.",
  "requestApprovalsRequestedStatus": true
}
```

Now this was an atypical payload for updating a DDFSIT SIT service item due to the following reasons:
1. We don't allow updates to DDFSIT service items
2. `reason` is not a valid property in the `updateMTOServiceItem` endpoint (this should be `updateReason`)
3. This looks a loooot like the payload that is used when CREATING (or requesting) destination SIT - which automatically creates the additional destination SIT service items when the DDFSIT service item is created

We had a bunch of meetings and it came to realization that they thought that, when resubmitting a rejected SIT service item, they only had to update the 1st Day SIT service item and not all of them. Once that was discovered, we promptly told them that we handle resubmitting rejected SIT service items at an individual level and there is no logic within that update endpoint that mirrors what happens when creating `DDFSIT` or `DOFSIT`.

We landed on a fix and this PR does the following:
- opens up the `updateMTOServiceItem` endpoint to also allow for `DDFSIT`, `DOSFSC`, and `DDSFSC` service item codes
- adds validator that checks to make sure that the `reason` value has changed from the previous SIT service item that was in a `REJECTED` status
- refactored the `mto_service_item_validators.go` file a bit and rearranged checks and error returns
- checking move `status` after updating Prime service items to analyze if it needs TOO review
- updated Prime API to add details when resubmitting rejected SIT service items
- added all relevant tests following changes of functionality
- random code cleanup I found when analyzing


### How to test

1. As the Prime, find a shipment that doesn't currently have origin or destination SIT
2. In the Prime Sim, request either origin/destination SIT and complete the form
3. As the TOO, go in and reject all of the newly added service items
4. Using Postman (sorry - gotta do it this way), using the endpoint 
`http://primelocal:3000/prime/v1/mto-service-items/{mtoServiceItemID}`
provide a payload that looks like this
```
{
    "updateReason": "a reason that is different than the previous reason",
    "requestApprovalsRequestedStatus": true,
    "modelType": "UpdateMTOServiceItemSIT",
    "reServiceCode": "DDFSIT"
}
```
5. You *should* get a successful response
6. Do this for each SIT service item
7. Try not updating the reason and ensure you get an error back
